File: demo/flatbuffer_input_data/builtin_prefix/read_fb_data.py
import sys
import argparse

sys.path.append("..")

import generated.VW.parsers.flatbuffer.Example as e
import generated.VW.parsers.flatbuffer.ExampleCollection as ec
import generated.VW.parsers.flatbuffer.ExampleRoot as er
import generated.VW.parsers.flatbuffer.ExampleType as et

def print_example(example):
    ns = example.Namespaces(0)
    print(f"namespace: {ns.Name()}")
    for j in range(ns.FeaturesLength()):
        feature = ns.Features(j)
        print(f"feature name: {feature.Name()}, feature value: {feature.Value()}")

def deserialize_and_print(from_file):
    with open(from_file, 'rb') as data_file:
        while True:
            # read first 32 bits holding the object size
            prefix = data_file.read(4)
            if len(prefix) <= 0:
                print("Done reading file")
                break
            prefix_obj_size = int.from_bytes(prefix, byteorder='little')

            # read the rest of the object to be deserialized
            buf = data_file.read(prefix_obj_size)
            buf = bytearray(buf)

            root_example = er.ExampleRoot.GetRootAsExampleRoot(buf, 0)
            tentative_object = root_example.ExampleObj()
            
            if root_example.ExampleObjType() == et.ExampleType().Example:
                example = e.Example()
                example.Init(tentative_object.Bytes, tentative_object.Pos)
                # examples currently have only one namespace
                print_example(example)
            elif root_example.ExampleObjType() == et.ExampleType().ExampleCollection:
                example_coll = ec.ExampleCollection()
                example_coll.Init(tentative_object.Bytes, tentative_object.Pos)
                coll_size = example_coll.ExamplesLength()
                for i in range(coll_size):
                    ex = example_coll.Examples(i)
                    print_example(example_coll.Examples(i))

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--from-file",
        help="flatbuffer file to deserialize",
        required=True,
    )

    args = parser.parse_args()
    deserialize_and_print(args.from_file)


if __name__ == "__main__":
    main()

File: demo/flatbuffer_input_data/builtin_prefix/read_fb_data.cc
#include "convert_to_fb.h"

void deserialize_and_print(const std::string& from)
{
  std::ifstream data_file(from, std::ios::binary | std::ios::in);
  size_t prefix_len = sizeof(uint32_t);
  uint32_t prefix;

  while (!data_file.fail())
  {
    // read the prefix from the file
    data_file.read((char*)&prefix, prefix_len);
    if (data_file.fail())
    {
      std::cout << "Done reading file" << std::endl;
      break;
    }

    std::vector<char> fb_obj(prefix);
    // read one object, object size defined by the read prefix
    data_file.read(&fb_obj[0], prefix);

    auto flatbuffer_pointer = reinterpret_cast<u_int8_t*>(&fb_obj[0]);

    auto data = GetExampleRoot(flatbuffer_pointer);
    switch (data->example_obj_type())
    {
      case ExampleType_Example:
      {
        const auto example = data->example_obj_as_Example();
        print_example(example);
      }
      break;
      case ExampleType_ExampleCollection:
      {
        const auto example_coll = data->example_obj_as_ExampleCollection();
        const auto example_size = example_coll->examples()->Length();
        for (size_t i = 0; i < example_size; i++)
        {
          const auto example = example_coll->examples()->Get(i);
          print_example(example);
        }
      }
      break;

      default:
        break;
    }
  }
}

int main(int argc, char* argv[])
{
  if (argc < 2) { throw std::invalid_argument("You need to provide a flatbuffer file to be deserialized"); }

  deserialize_and_print(argv[1]);
  return 0;
}